d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/PartitionState.java

```diff
@@ -0,0 +1,249 @@
+import com.linkedin.d2.balancer.strategies.degrader.LoadBalancerQuarantine;
+import com.linkedin.d2.balancer.strategies.degrader.RingFactory;
+import com.linkedin.d2.balancer.util.hashing.Ring;
+import com.linkedin.d2.balancer.util.healthcheck.HealthCheck;
+import com.linkedin.util.degrader.CallTracker;
+import java.net.URI;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+
+/**
+ * Each {@link PartitionState} corresponds to a partition for a particular service in the relative load balancer
+ * It keeps the tracker clients and the ring of the partition
+ */
+public class PartitionState
+{
+  private final int _partitionId;
+  private final int _pointsPerWeight;
+  private final RingFactory<URI> _ringFactory;
+  private final List<PartitionStateUpdateListener<PartitionState>> _listeners;
+  private Set<TrackerClient> _recoveryTrackerClients;
+  private long _clusterGenerationId;
+  private Map<TrackerClient, LoadBalancerQuarantine> _quarantineMap;
+  private Map<TrackerClient, LoadBalancerQuarantine> _quarantineHistory;
+  private Map<TrackerClient, HealthCheck> _healthCheckMap;
+  private Map<URI, Integer> _pointsMap;
+  private Ring<URI> _ring;
+  private Map<TrackerClient, TrackerClientState> _trackerClientStateMap;
+  private PartitionStats _partitionStats;
+
+  public PartitionState(int partitionId, RingFactory<URI> ringFactory, int pointsPerWeight,
+      List<PartitionStateUpdateListener<PartitionState>> listeners)
+  {
+    _partitionId = partitionId;
+    _clusterGenerationId = -1;
+    _ringFactory = ringFactory;
+    _pointsPerWeight = pointsPerWeight;
+
+    _recoveryTrackerClients = new HashSet<>();
+    _quarantineMap = new HashMap<>();
+    _quarantineHistory = new HashMap<>();
+    _healthCheckMap = new HashMap<>();
+    _trackerClientStateMap = new HashMap<>();
+    _listeners = listeners;
+    resetRing();
+  }
+
+  PartitionState(int partitionId, RingFactory<URI> ringFactory, int pointsPerWeight,
+      Set<TrackerClient> recoveryTrackerClients, long clusterGenerationId,
+      Map<TrackerClient, LoadBalancerQuarantine> quarantineMap,
+      Map<TrackerClient, LoadBalancerQuarantine> quarantineHistory,
+      Map<TrackerClient, HealthCheck> healthCheckMap,
+      Map<TrackerClient, TrackerClientState> trackerClientStateMap,
+      List<PartitionStateUpdateListener<PartitionState>> listeners)
+  {
+    _partitionId = partitionId;
+    _ringFactory = ringFactory;
+    _pointsPerWeight = pointsPerWeight;
+    _recoveryTrackerClients = recoveryTrackerClients;
+    _clusterGenerationId = clusterGenerationId;
+    _quarantineMap = quarantineMap;
+    _quarantineHistory = quarantineHistory;
+    _healthCheckMap = healthCheckMap;
+    _trackerClientStateMap = trackerClientStateMap;
+    _listeners = listeners;
+    resetRing();
+  }
+
+  public PartitionState copy()
```

Comment:
It's not purely shallowCopy, especially for the Map variables. I created a new map for the new state, so that if we add new elements or remove outdated old elements in the new state, it will not change the old state. Want to keep the old state immutable.

Reply:
Trying to come up a new name because copy/clone suggest reuse the objects. Maybe constructNew or build?

Reply:
I changed it to a constructor instead, does this work better?
public PartitionState (PartitionState oldPartitionState)
  {
    this(oldPartitionState.getPartitionId(),
        oldPartitionState.getRingFactory(),
        oldPartitionState.getPointsPerWeight(),
        new HashSet<>(oldPartitionState.getRecoveryTrackerClients()),
        oldPartitionState.getClusterGenerationId(),
        new HashMap<>(oldPartitionState.getQuarantineMap()),
        new HashMap<>(oldPartitionState.getQuarantineHistory()),
        new HashMap<>(oldPartitionState.getHealthCheckMap()),
        new HashMap<>(oldPartitionState.getTrackerClientStateMap()),
        oldPartitionState.getListeners());
  }

Reply:
Yes!